htmd/parameterization/ffmolecule.py

```diff
@@ -261,518 +243,117 @@ def getDipole(self):
+        return [dihedral.atoms.copy() for dihedral in self._soft_dihedrals]
 
-        self._makeDihedralUnique(phi_to_fit)
+    def fitSoftTorsions(self, dihedrals, geomopt=True):
+        """
+        Dihedrals passed as 4 atom indices
+        """
 
-        atoms = phi_to_fit.atoms
-        equivs = phi_to_fit.equivalents
+        # Create molecules with romtamers
+        molecules = []
+        for dihedral in dihedrals:
 
-        # Number of rotamers for each dihedral to compute
-        nrotamer = 36
+            nrotamers = 36  # Number of rotamers for each dihedral to compute
 
-        # Create a copy of molecule with nrotamer frames
-        mol = self.copy()
-        for _ in range(nrotamer-1):
-            mol.appendFrames(self)
-        assert mol.numFrames == nrotamer
+            # Create a copy of molecule with "nrotamers" frames
+            mol = self.copy()
+            while mol.numFrames < nrotamers:
+                mol.appendFrames(self)
+            assert mol.numFrames == nrotamers
 
-        # Set rotamer coordinates
-        angles = np.linspace(-np.pi, np.pi, num=nrotamer, endpoint=False)
-        for frame, angle in enumerate(angles):
-            mol.frame = frame
-            mol.setDihedral(atoms, angle, bonds=mol.bonds)
+            # Set rotamer coordinates
+            angles = np.linspace(-np.pi, np.pi, num=nrotamers, endpoint=False)
+            for frame, angle in enumerate(angles):
+                mol.frame = frame
+                mol.setDihedral(dihedral, angle, bonds=mol.bonds)
 
-        dirname = 'dihedral-opt' if geomopt else 'dihedral-single-point'
-        dih_name = "%s-%s-%s-%s" % (self.name[atoms[0]], self.name[atoms[1]], self.name[atoms[2]], self.name[atoms[3]])
-        fitdir = os.path.join(self.outdir, dirname, dih_name, self.output_directory_name())
+            molecules.append(mol)
 
-        try:
+        # Run QM calculation of the romaters
+        dirname = 'dihedral-opt' if geomopt else 'dihedral-single-point'
+        qm_results = []
+        for dihedral, molecule in zip(dihedrals, molecules):
+            name = "%s-%s-%s-%s" % tuple(self.name[dihedral])
+            fitdir = os.path.join(self.outdir, dirname, name, self.output_directory_name())
             os.makedirs(fitdir, exist_ok=True)
-        except:
-            raise OSError('Directory {} could not be created. Check if you have permissions.'.format(fitdir))
 
-        self.qm.molecule = mol
-        self.qm.esp_points = None
-        self.qm.optimize = geomopt
-        self.qm.restrained_dihedrals = np.array(frozens)
-        self.qm.directory = fitdir
-        results = self.qm.run()
+            self.qm.molecule = molecule
+            self.qm.esp_points = None
+            self.qm.optimize = geomopt
+            self.qm.restrained_dihedrals = np.array([dihedral])
+            self.qm.directory = fitdir
+            qm_results.append(self.qm.run())  # TODO submit all jobs at once
```

Comment:
Doesn't this mean that each rotamer is done in a serial way? This needs to be changed before merging.

Reply:
No! QM class already parallelize over rotamers. The remaining issue, it is not parallized over dihedrals, i.e. QM jobs will be run in batches of 36, but not all at once.

Until AceCloud is available, we can live with that.

Reply:
ok ok